src/cpp/ext/filters/census/environment_autodetect.cc

```diff
@@ -0,0 +1,349 @@
+}
+
+// Get service name for Cloud run
+std::string GetServiceName() {
+  return grpc_core::GetEnv("K_SERVICE").value_or("");
+}
+
+// Get configuration name for Cloud run
+std::string GetConfiguratioName() {
+  return grpc_core::GetEnv("K_CONFIGURATION").value_or("");
+}
+
+// Get module ID for App Engine
+std::string GetModuleId() {
+  return grpc_core::GetEnv("GAE_SERVICE").value_or("");
+}
+
+// Get version ID for App Engine
+std::string GetVersionId() {
+  return grpc_core::GetEnv("GAE_VERSION").value_or("");
+}
+
+// Fire and forget class
+class EnvironmentAutoDetectHelper
+    : public grpc_core::InternallyRefCounted<EnvironmentAutoDetectHelper>,
+      private internal::GrpcLibrary {
+ public:
+  explicit EnvironmentAutoDetectHelper(
+      std::string project_id,
+      absl::AnyInvocable<void(EnvironmentAutoDetect::ResourceType)> on_done,
+      std::shared_ptr<grpc_event_engine::experimental::EventEngine>
+          event_engine)
+      : InternallyRefCounted(/*trace=*/nullptr, /*initial_refcount=*/2),
+        project_id_(std::move(project_id)),
+        on_done_(std::move(on_done)),
+        event_engine_(std::move(event_engine)) {
+    grpc_core::ExecCtx exec_ctx;
+    // TODO(yashykt): The pollset stuff should go away once the HTTP library is
+    // ported over to use EventEngine.
+    pollset_ = static_cast<grpc_pollset*>(gpr_zalloc(grpc_pollset_size()));
+    grpc_pollset_init(pollset_, &mu_poll_);
+    pollent_ = grpc_polling_entity_create_from_pollset(pollset_);
+    // TODO(yashykt): Note that using EventEngine::Run is not fork-safe. If we
+    // want to make this fork-safe, we might need some re-work here.
+    event_engine_->Run([this] { PollLoop(); });
+    AutoDetect();
+  }
+
+  ~EnvironmentAutoDetectHelper() override {
+    grpc_core::ExecCtx exec_ctx;
+    grpc_pollset_shutdown(
+        pollset_, GRPC_CLOSURE_CREATE(
+                      [](void* arg, absl::Status /* status */) {
+                        grpc_pollset_destroy(static_cast<grpc_pollset*>(arg));
+                        gpr_free(arg);
+                      },
+                      pollset_, nullptr));
+  }
+
+  void Orphan() override {
+    grpc_core::Crash("Illegal Orphan() call on EnvironmentAutoDetectHelper.");
+  }
+
+ private:
+  struct Attribute {
+    std::string resource_attribute;
+    std::string metadata_server_atttribute;
+  };
+
+  void PollLoop() {
+    gpr_log(GPR_ERROR, "polling");
+    grpc_core::ExecCtx exec_ctx;
+    bool done = false;
+    gpr_mu_lock(mu_poll_);
+    grpc_pollset_worker* worker = nullptr;
+    if (!GRPC_LOG_IF_ERROR(
+            "pollset_work",
+            grpc_pollset_work(grpc_polling_entity_pollset(&pollent_), &worker,
+                              grpc_core::Timestamp::InfPast()))) {
+      notify_poller_ = true;
+    }
+    done = notify_poller_;
+    gpr_mu_unlock(mu_poll_);
+    if (!done) {
+      event_engine_->RunAfter(grpc_core::Duration::Milliseconds(100),
```

Comment:
The performance hit is probably negligible, but I wonder if this would be better written as a loop with a 100ms sleep, than a lambda-wrapped recursive function with a thread hop.

```
bool done = false;
while (!done) {
  // ...
}
Unref();
```

Reply:
Isn't it bad to do a blocking sleep in an event manager thread. That seems like something we definitely do not want to do. Also note that in this case, we really do not care about performance.